### PR TITLE
lint aot_autograd with mypy

### PR DESCRIPTION
**Pull request by bdhirsh**
I added a handful of ignores, but the majority of aot_autograd.py now passes with mypy (which fixed a few minor bugs)

I'm not 100% sold on whether or not this is useful (I have a bunch of mypy errors to fix in my AOTDispatch PR). Happy to land if people like to the type checking, or otherwise toss.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #97540
* __->__ #96784

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

File: torch/_functorch/aot_autograd.py [hunk 1]
```diff
@@ -2929,17 +2952,19 @@ def aot_module_simplified(
...
     }
-    params_flat, params_spec = pytree.tree_flatten(params)
-    params_flat = tuple(params_flat)
+    params_flat_, params_spec = pytree.tree_flatten(params)
+    params_flat: Tuple[Any, ...] = params_flat_ if isinstance(params_flat_, tuple) else tuple(params_flat_)
     params_len = len(params_flat)
 
     def functional_call(*args, **kwargs):
+        non_param_args = args[:params_len]
+        assert isinstance(non_param_args, list)
```
Comment (ezyang) [inline torch/_functorch/aot_autograd.py:2965]: ??? this isn't even true

```
>>> (1,2,3)[2:]
(3,)
```

File: torch/_functorch/aot_autograd.py [hunk 2]
```diff
@@ -1622,22 +1626,29 @@ def create_synthetic_base_metadata(
     output_grad_info = m.requires_grad_info[num_original_input_data_mutations:]
     input_metadata_mutation_grad_info = [
         outer_args[outer_idx].requires_grad for outer_idx in outer_aliased_arg_idx_with_metadata_mutations]
-    input_metadata_output_info = [
-        OutputAliasInfo(
+    input_metadata_output_info = []
```
Comment (ezyang) [inline torch/_functorch/aot_autograd.py:1629]: don't you have to annotate this

File: torch/_functorch/aot_autograd.py [hunk 3]
```diff
@@ -1622,22 +1626,29 @@ def create_synthetic_base_metadata(
...
+    for outer_idx in outer_aliased_arg_idx_with_metadata_mutations:
+        curr_info = synthetic_base_info[outer_idx]
+        assert isinstance(curr_info, tuple)
```
Comment (ezyang) [inline torch/_functorch/aot_autograd.py:1632]: why don't you just annotate `synthetic_base_info` properly

**Review by ezyang — CHANGES_REQUESTED** [review]
there seems to be a lot of unidiomatic fixes to type errors

**Comment by ezyang** [comment]
explicitly retag when you want review

**Pull request closed without merging**